bookwyrm/models/bookwyrm_export_job.py

```diff
@@ -0,0 +1,217 @@
+
+        return self
+
+
+@app.task(queue=IMPORTS, base=ParentTask)
+def start_export_task(**kwargs):
+    """trigger the child tasks for each row"""
+    job = BookwyrmExportJob.objects.get(id=kwargs["job_id"])
+
+    # don't start the job if it was stopped from the UI
+    if job.complete:
+        return
+    try:
+        # This is where ChildJobs get made
+        job.export_data = ContentFile(b"", str(uuid4()))
+        json_data = json_export(job.user)
+        tar_export(json_data, job.user, job.export_data)
+        job.save(update_fields=["export_data"])
+    except Exception as err:  # pylint: disable=broad-except
+        logger.exception("User Export Job %s Failed with error: %s", job.id, err)
+        job.set_status("failed")
+
+    job.set_status("complete")
+
+
+def tar_export(json_data: str, user, file):
+    """wrap the export information in a tar file"""
+    file.open("wb")
+    with BookwyrmTarFile.open(mode="w:gz", fileobj=file) as tar:
+        tar.write_bytes(json_data.encode("utf-8"))
+
+        # Add avatar image if present
+        if getattr(user, "avatar", False):
+            tar.add_image(user.avatar, filename="avatar")
+
+        editions, books = get_books_for_user(user)  # pylint: disable=unused-variable
+        for book in editions:
+            if getattr(book, "cover", False):
+                tar.add_image(book.cover)
+
+    file.close()
+
+
+def json_export(user):  # pylint: disable=too-many-locals, too-many-statements
+    """Generate an export for a user"""
+    # user
+    exported_user = {}
+    vals = [
+        "username",
+        "name",
+        "summary",
+        "manually_approves_followers",
+        "hide_follows",
+        "show_goal",
+        "show_suggested_users",
+        "discoverable",
+        "preferred_timezone",
+        "default_post_privacy",
+    ]
+    for k in vals:
+        exported_user[k] = getattr(user, k)
+
+    if getattr(user, "avatar", False):
+        exported_user["avatar"] = f'https://{DOMAIN}{getattr(user, "avatar").url}'
+
+    # reading goals
+    reading_goals = models.AnnualGoal.objects.filter(user=user).distinct()
+    goals_list = []
+    # TODO: either error checking should be more sophisticated
+    # or maybe we don't need this try/except
+    try:
+        for goal in reading_goals:
+            goals_list.append(
+                {"goal": goal.goal, "year": goal.year, "privacy": goal.privacy}
+            )
+    except Exception:  # pylint: disable=broad-except
+        pass
+
+    try:
+        readthroughs = models.ReadThrough.objects.filter(user=user).distinct().values()
+        readthroughs = list(readthroughs)
+    except Exception:  # pylint: disable=broad-except
+        readthroughs = []
+
+    # books
+    editions, books = get_books_for_user(user)
+    final_books = []
+
+    for book in books.values():
+        edition = editions.filter(id=book["id"])
+        book["edition"] = edition.values()[0]
```

Comment:
It took me a second to understand this, but it looks like you're loading the object as a queryset, and then using `queryset.values()` to give you an object representation of the edition, is that right?

I've been wondering if it's possible to use the activitypub formulation of books and authors, because then you could use the pre-existing code that converts activitypub editions/authors into model objects (that code also does deduplication against existing model entries). I think the only meaningful field related to the book itself that you'd miss was the cover? But I'm not sure if it would be prohibitively hard to relate the user's data (like readthrough) to the activitypub serialization.

Reply:
> It took me a second to understand this, but it looks like you're loading the object as a queryset, and then using queryset.values() to give you an object representation of the edition, is that right?

Correct.

I'd be very happy to adjust this with advice. Initially I thought we probably should do exports and imports as ActivityPub objects for the reason you've noted - most of the code is already there - but it was proving to be quite challenging. Then I was trying to do something kinda silly to get the Django model instances out as JSON, and then we ended up doing it this way. Images end up kind of having to be pulled out separately anyway because all we can get at this point is a filepath, so if you can walk me through what you're thinking a bit more, definitely happy to keep it simpler.

Reply:
I'm not surprised that using activitypub ended up having a lot of challenges, and I'm satisfied knowing that you tried it and it turned out to be a major headache.